viewing_party/party.py

```diff
@@ -1,23 +1,178 @@
+
+
+# 1st function in Wave 3
+def get_unique_watched(user_data):
+
+    friends_watched_list = []
+    friends_watched_set = None
+    friends_list = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_list.append(movie["title"])
+            friends_watched_set = set(friends_watched_list)
+
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_watched_set and movie not in friends_list:
+            friends_list.append(movie)
+
+    return friends_list
+
+
+# 2nd function in Wave 3
+def get_friends_unique_watched(user_data):
+
+    user_watched = set([movie["title"] for movie in user_data["watched"]])
+    friends_list = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched and movie not in friends_list:
+                friends_list.append(movie)
+
+    return friends_list
+
+
+# 1st function in Wave 4
+def get_available_recs(user_data):
+
+    unseen_by_user = get_friends_unique_watched(user_data)
+    user_subscriptions = user_data["subscriptions"]
+    available_recs = []
+
+    for movie in unseen_by_user:
+        if movie["host"] in user_subscriptions:
+            available_recs.append(movie)
+    return available_recs
+
+
+#1st function Wave 5
+def get_new_rec_by_genre(user_data):
+
+    user_watched = set([movie["title"] for movie in user_data["watched"]])
+    friends_unique_watched = get_friends_unique_watched(user_data)
+
+    genre_count = {}
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        genre_count[genre] = genre_count.get(genre, 0) + 1
+    best_genre = None
+    best_genre_count = 0
+    for genre, count in genre_count.items():
+        if count > best_genre_count:
+            best_genre = genre
+            best_genre_count = count
 
+    recommended_movies = []
+    for movie in friends_unique_watched:
+        if movie["title"] not in user_watched and movie["genre"] == best_genre:
+            recommended_movies.append(movie)
 
-# -----------------------------------------
-# ------------- WAVE 3 --------------------
-# -----------------------------------------
+    return recommended_movies
 
-        
-# -----------------------------------------
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
 
-# -----------------------------------------
-# ------------- WAVE 5 --------------------
-# -----------------------------------------
+# 2ND function in Wave 5
+def get_rec_from_favorites(user_data):
+    
+    user_favorites = user_data["favorites"]
+    recommended_movies = []
 
+    if user_data["friends"] == []:
+        return user_favorites
+    else:
+        unseen_by_friends = get_unique_watched(user_data)
+        for movie in unseen_by_friends:
+            if movie in user_favorites:
+                recommended_movies.append(movie)
+        return recommended_movies
```

Comment:
Here, your code checks if `user_data` has friends. If the user has friends, then it iterates through the `unseen_by_friends` list. If the user doesn't have friends, then the function should return an empty list, as there are no recommendations.

However, when we iterate with a `for` loop over an empty list, the code shouldn't break. There's a way to get rid of this `if`/`else`, but it would require fixing a small bug in `get_unique_watched`. Consider the following:

Imagine that `get_unique_watched` returned an empty list. If `unseen_by_friends` is an empty list, then when we do `for movie in unseen_by_friends`, it will skip over this loop, and `recommended_movies` will be an empty list. Therefore, with the fix in `get_unique_watched`, this function could look like this:

```python
def get_rec_from_favorites(user_data):
    
    user_favorites = user_data["favorites"]
    recommended_movies = []

    unseen_by_friends = get_unique_watched(user_data)
    for movie in unseen_by_friends:
        if movie in user_favorites:
            recommended_movies.append(movie)
    return recommended_movies
```

I got this version of the code to work if, inside `get_unique_watched`, I moved the line `friends_watched_set = set(friends_watched_list)`. Refer to that comment for details!

The reason why I like this refactoring is because we get to reduce the amount of validation, and we don't need to repeat it over several functions. Let me know what questions come up!